test/binaryen.js/kitchen-sink.js

```diff
@@ -178,6 +208,9 @@ function test_core() {
     module.unreachable(),
   ];
 
+  // Test expression utility
+  console.log("ExpressionGetId=" + module.getExpressionId(valueList[3]));
```

Comment:
another option is to put these two methods on `Binaryen`, as they don't need a module to work. So `Binaryen.getExpressionId(..)`. Thoughts?

Reply:
Yeah, though about this as well. Could do either, or both. Ultimately, when thinking about it, one thing that came to my mind was to return instances of a new class  `Expression`, holding a `BinaryenExpressionRef`, with these expression methods, including `print`. That'd be quite similar to `Module` and `Relooper`, somewhat connected to #1273 and not yet doable without other functions working with references (i.e. could do that for everything, like `Global`s, `Function`s, `FunctionType`s ...).

Reply:
I don't have a clear preference there because I have switched to using the C-API directly from both JS and WASM, hoping that it becomes interchangeable eventually ([see](https://github.com/dcodeIO/assemblyscript-next/blob/master/src/binaryen.d.ts)). Not sure if that's an option for upstream as well in the long them, i.e. by making the wrapper just an npm package instead of compiling it in (generating just the raw C-API compiled to JS that can be used as a dependency for such wrappers).

Reply:
Returning a JS class for every expression would be more JS-ey, yeah, but I think it's better not to for perf reasons: we have just one (or a few) Module and Relooper instances, but potentially a huge number of expressions, so my intention was to avoid JS allocation for them.

Not sure I understand your second comment, and first link there is a 404?

Reply:
Yeah, from a performance perspective it might be better to stick to expression references. Considering performance, it might also be better to use proper prototypes instead of assigning instance methods through `this['emitText'] =` etc. as well.

My idea above was (sorry, updated the links to point to the public repo) to provide just the bare minimum (the raw C-API compiled to JS/WASM) by WebAssembly/binaryen, leaving the respective wrappers, like one using just references or one using classes for everything, up to external modules because there might be more than one wrapper in JS-space anyway, looking at all the languages that transpile to it.

Reply:
Prototypes might be more standard JS, yeah. Could also help perf, not sure.

Why do you prefer the raw C API exposed to JS instead of the more JS-ey one? If that's more convenient it would be less work, too, maybe I was wrong that we need more.

Reply:
Might be that my requirements are a bit different. Basically, I am speculating on the feasibility of compiling Binaryen to WASM one day, and then being able to simply switch between the two (either calling from JS or linking into a WASM module) with the only common denominator being the raw C-API. In the AssemblyScript case this makes even more sense because the higher level wrapper (2nd link) can be shared between TS and AS as well.

Regarding just providing a minimal API: As I know the crazy people doing npm, there'd probably be multiple wrappers around in a week or two, each focusing on other things (i.e. one being entirely class-based, one being just a few helpers to the C-API and another being written in CoffeScript or something). Not super important, though, because binaryen.js, as it stands, still exposes the underlying raw C-API. Just thinking aloud :)

Reply:
Got it, thanks for the info.